src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            {
+                return null;
+            }
+
+            var foreachVariable = operation.Locals[0];
+            if (foreachVariable == null)
+            {
+                return null;
+            }
+
+            if (!CheckForeachVariable(model, foreachVariable, foreachStatement))
+            {
+                return null;
+            }
+
+            var foreachCollection = RemoveImplicitConversion(operation.Collection);
+            if (foreachCollection == null)
+            {
+                return null;
+            }
+
+            string explicitCast;
+            string countName;
+            GetInterfaceInfo(semanticFact, model, foreachVariable, foreachCollection, out explicitCast, out countName);
+            if (countName == null)
+            {
+                return null;
+            }
+
+            var requireCollectionStatement = CheckRequireCollectionStatement(foreachCollection);
+            return new ForeachInfo(countName, explicitCast, requireCollectionStatement, foreachStatement);
+        }
+
+        private static void GetInterfaceInfo(
+            ISemanticFactsService semanticFact, SemanticModel model, ILocalSymbol foreachVariable, IOperation foreachCollection, out string explicitCast, out string countName)
+        {
+            explicitCast = default;
+            countName = null;
+
+            // go through list of types to find out right set;
+            var foreachType = foreachVariable.Type;
+            if (IsNullOrErrorType(foreachType))
+            {
+                return;
+            }
+
+            // check array case first.
+            var collectionType = foreachCollection.Type;
+            if (IsNullOrErrorType(collectionType))
+            {
+                return;
+            }
+
+            if (collectionType is IArrayTypeSymbol array &&
+                array.Rank == 1 &&
+                semanticFact.IsAssignableTo(array.ElementType, foreachType, model.Compilation))
+            {
+                explicitCast = null;
+                countName = Length;
+                return;
+            }
+
+            // check ImmutableArray case second
+            if (collectionType.OriginalDefinition.Equals(model.Compilation.GetTypeByMetadataName(typeof(ImmutableArray<>).FullName)))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCast = null;
+                    countName = Length;
+                    return;
+                }
+            }
+
+            // go through all known interfaces we support
+            var knownCollectionInterfaces = s_KnownInterfaceNames.Select(
+                s => model.Compilation.GetTypeByMetadataName(s)).WhereNotNull().Where(t => !IsNullOrErrorType(t));
```

Comment:
you already did WhereNotNull.  So why do another IsNull check?

Reply:
I added that for ErrorType. yep. WhereNotNull is not needed.